src/Sarif/Core/Run.cs

```diff
@@ -138,7 +140,37 @@ private void InitializeFileToIndexMap()
         /// <returns>ToolComponent for reference</returns>
         public ToolComponent GetToolComponentFromReference(ToolComponentReference reference)
         {
-            return this.Tool.GetToolComponentFromReference(reference);
+            return this.Tool?.GetToolComponentFromReference(reference);
+        }
+
+        /// <summary>
+        ///  Set the Run property on each Result to this Run, so that Result methods
+        ///  and properties which may need to look up Run collections can do so.
+        /// </summary>
+        public void SetRunOnResults()
+        {
+            if (this.Results != null)
+            {
+                DeferredList<Result> deferredResults = this.Results as DeferredList<Result>;
```

Comment:
>DeferredList [](start = 16, length = 12)

Oh boy. This pattern makes me wonder whether we need all Lists in the format to be able to distinguish between these conditions? Otherwise we're on a journey to endlessly chase down bugs in various iteration scenarios. Not relevant to this check-in.

Reply:
Yes, this part was pretty lame.

*If* the deserialization of Results could automatically pick up on the Run, the Deferred infrastructure would probably "just work" without this special stuff.

Otherwise, anywhere we have an IList<T> which may be deferred and we have an "outer object reference setter" like this, we need the same pattern.

Maybe we could figure out how to design this so that we can load a comprehensive log with the deferred OM and confirm in one go that all of the upward object references were properly set.

Reply:
sounds good.

---
In reply to: [330692856](https://github.com/microsoft/sarif-sdk/pull/1691#discussion_r330692856) [](ancestors = 330692856)